pkg/queryfrontend/labels_codec_test.go

```diff
@@ -521,3 +522,209 @@ func TestLabelsCodec_MergeResponse(t *testing.T) {
+		})
+		testutil.Ok(b, err)
+
+		b.ReportAllocs()
+		b.ResetTimer()
+
+		for n := 0; n < b.N; n++ {
+			_, err := codec.DecodeResponse(
+				ctx,
+				makeResponse(seriesData, false),
+				&ThanosSeriesRequest{})
+			testutil.Ok(b, err)
+		}
+	})
+
+	b.Run("SeriesResponseWithHeaders", func(b *testing.B) {
+		seriesDataWithHeaders, err := json.Marshal(&ThanosSeriesResponse{
+			Status:  "success",
+			Data:    []labelpb.ZLabelSet{{Labels: []labelpb.ZLabel{{Name: "foo", Value: "bar"}}}},
+			Headers: []*ResponseHeader{{Name: cacheControlHeader, Values: []string{noStoreValue}}},
+		})
+		testutil.Ok(b, err)
+
+		b.ReportAllocs()
+		b.ResetTimer()
+
+		for n := 0; n < b.N; n++ {
+			_, err := codec.DecodeResponse(
+				ctx,
+				makeResponse(seriesDataWithHeaders, true),
+				&ThanosSeriesRequest{})
+			testutil.Ok(b, err)
+		}
+	})
+
+	b.Run("LabelsResponse", func(b *testing.B) {
+		labelsData, err := json.Marshal(&ThanosLabelsResponse{
+			Status: "success",
+			Data:   []string{"__name__"},
+		})
+		testutil.Ok(b, err)
+
+		b.ReportAllocs()
+		b.ResetTimer()
+
+		for n := 0; n < b.N; n++ {
+			_, err := codec.DecodeResponse(
+				ctx,
+				makeResponse(labelsData, false),
+				&ThanosLabelsRequest{})
+			testutil.Ok(b, err)
+		}
+	})
+
+	b.Run("LabelsResponseWithHeaders", func(b *testing.B) {
+		labelsDataWithHeaders, err := json.Marshal(&ThanosLabelsResponse{
+			Status:  "success",
+			Data:    []string{"__name__"},
+			Headers: []*ResponseHeader{{Name: cacheControlHeader, Values: []string{noStoreValue}}},
+		})
+		testutil.Ok(b, err)
+
+		b.ReportAllocs()
+		b.ResetTimer()
+
+		for n := 0; n < b.N; n++ {
+			_, err := codec.DecodeResponse(
+				ctx,
+				makeResponse(labelsDataWithHeaders, true),
+				&ThanosLabelsRequest{})
+			testutil.Ok(b, err)
+		}
+	})
+}
+
+func BenchmarkLabelsCodecMergeResponses_1(b *testing.B) {
+	benchmarkMergeResponses(b, 1)
+}
+
+func BenchmarkLabelsCodecMergeResponses_10(b *testing.B) {
+	benchmarkMergeResponses(b, 10)
+}
+
+func BenchmarkLabelsCodecMergeResponses_100(b *testing.B) {
+	benchmarkMergeResponses(b, 100)
+}
+
+func BenchmarkLabelsCodecMergeResponses_1000(b *testing.B) {
+	benchmarkMergeResponses(b, 1000)
+}
+
+func benchmarkMergeResponses(b *testing.B, size int) {
+	codec := NewThanosLabelsCodec(false, time.Hour*2)
+	queryResLabel, queryResSeries := makeQueryRangeResponses(size)
+
+	b.Run("SeriesResponses", func(b *testing.B) {
+		b.ReportAllocs()
+		b.ResetTimer()
+
+		for i := 0; i < b.N; i++ {
+			_, _ = codec.MergeResponse(queryResSeries...)
+		}
+	})
+
+	b.Run("LabelsResponses", func(b *testing.B) {
+		b.ReportAllocs()
+		b.ResetTimer()
+
+		for i := 0; i < b.N; i++ {
+			_, _ = codec.MergeResponse(queryResLabel...)
+		}
+	})
+
+}
+
+func makeQueryRangeResponses(size int) ([]queryrange.Response, []queryrange.Response) {
+	labelResp := make([]queryrange.Response, 0, size)
+	seriesResp := make([]queryrange.Response, 0, size)
```

Comment:
Does this need to be `size*2` since we are generating 2 elements in each iteration of the loop?